agent/consul/rpc.go

```diff
@@ -1029,17 +1045,23 @@ func (s *Server) blockingQuery(
 
 		err := query(ws, state)
 		s.setQueryMeta(responseMeta, opts.GetToken())
+
 		switch {
 		case errors.Is(err, errNotFound):
 			if notFound {
 				// query result has not changed
 				minQueryIndex = responseMeta.GetIndex()
 			}
-
 			notFound = true
+		case errors.Is(err, errNotChanged):
+			if ranOnce {
+				// query result has not changed
+				minQueryIndex = responseMeta.GetIndex()
+			}
```

Comment:
I haven't thought too much about this yet, but I'm wondering if `blockingQuery` needs to be aware of this.

Could the closure set `responseMeta.Index` to the `query.Index` value itself? That might allow the closure to do this very same thing without needing `blockingQuery` to be aware of it.  Maybe there are cases I'm missing.